core/error_utils.go

```diff
@@ -0,0 +1,188 @@
+package core
+
+// (C) Copyright IBM Corp. 2024.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import (
+	"crypto/sha256"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"runtime"
+	"strings"
+
+	"github.com/go-yaml/yaml"
+)
+
+// Private utility functions for our custom error system
+
+// CreateIDHash computes a unique ID based on a given prefix
+// and problem attribute fields.
+func CreateIDHash(prefix string, fields ...string) string {
+	signature := strings.Join(fields, "")
+	hash := sha256.Sum256([]byte(signature))
+	return fmt.Sprintf("%s_%s", prefix, hex.EncodeToString(hash[:4]))
+}
+
+// getPreviousProblemID returns the ID of the "causedBy" problem, if it exists.
+func getPreviousProblemID(problem Problem) string {
+	if problem != nil {
+		return problem.GetID()
+	}
+	return ""
+}
+
+// getComponentInfo is a convenient way to access the name of the
+// component alongside the current semantic version of the library.
+func getComponentInfo() (string, string) {
+	return "github.com/IBM/go-sdk-core/v5", __VERSION__
+}
+
+// computeFunctionName investigates the program counter at a fixed
+// skip number (aka point in the stack) of 2, which gives us the
+// information about the function the problem was created in, and
+// returns the name of the function.
+func computeFunctionName(component string) string {
+	if pc, _, _, ok := runtime.Caller(2); ok {
+		// The function names will have the component as a prefix - to avoid
+		// redundancy, since we are including the component string with the
+		// problem, trim that prefix here.
+		return strings.TrimPrefix(runtime.FuncForPC(pc).Name(), component+"/")
+	}
+
+	return ""
+}
+
+// sdkStackFrame is a convenience struct for formatting
+// frame data to be printed as YAML.
+type sdkStackFrame struct {
+	Function string `json:"function,omitempty"`
```

Comment:
Should we have `json` tags on these fields or `yaml` tags if we're going to be doing yaml serialization?

Reply:
Good call. Actually, with the YAML package I ended up using, the keys are "lowercased" automatically when marshalling so we don't need them at all. I will remove them.